### PR TITLE
Split staking basic logic to a KernelStakes separate contract

### PR DESCRIPTION
**Pull request by spalladino**
Move staking storage and basic logic to a separate contract. Tests pending.

### CODE REVIEW

File: kernel/contracts/ZepCore.sol [hunk 1]
```diff
@@ -3,77 +3,60 @@ pragma solidity ^0.4.18;
...
   }
 
-  function totalStakedFor(KernelInstance instance) public view returns (uint256) {
-    return _instanceVouches[instance];
-  }
-
-  function totalStaked() public view returns (uint256) {
-    return _totalStaked;
+  function transferStake(KernelInstance from, KernelInstance to, uint256 amount, bytes data) public {
+    _stakes.transferStake(msg.sender, from, to, amount, data);
```
Comment (fiiiu) [inline kernel/contracts/ZepCore.sol]: Oh! But the developer of the `to` KernelInstance is not getting her share of tokens, is she?

**Comment by fiiiu** [comment]
Addresses #61.

File: kernel/contracts/KernelStakes.sol [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    _totalStaked = totalStaked().sub(amount);
+    _instanceVouches[instance] = _instanceVouches[instance].sub(amount);
+    _userVouches[staker][instance] = currentStake.sub(amount);
+    
+    Unstaked(staker, instance, amount, _instanceVouches[instance], data);
+  }
+
+  function transferStake(address staker, address from, address to, uint256 amount, bytes data) public onlyOwner {
+    unstake(staker, from, amount, data);
+    stake(staker, to, amount, data);
```
Comment (facuspagnuolo) [inline kernel/contracts/KernelStakes.sol]: we should send the payout to the new dev here

File: kernel/contracts/KernelStakes.sol [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  function totalStakedFor(address instance) public view returns (uint256) {
+    return _instanceVouches[instance];
+  }
+
+  function totalStaked() public view returns (uint256) {
+    return _totalStaked;
+  }
+}
```
Comment (facuspagnuolo) [inline kernel/contracts/KernelStakes.sol]: Love how this looks like 👍

File: kernel/contracts/KernelStakes.sol [hunk 4]
```diff
@@ -0,0 +1,55 @@
+pragma solidity ^0.4.18;
+
+import "./ZepToken.sol";
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+import "zeppelin-solidity/contracts/ownership/Ownable.sol";
+
+contract KernelStakes is  Ownable {
```
Comment (facuspagnuolo) [inline kernel/contracts/KernelStakes.sol]: Just an idea, given this contract is the one who process staking and unstaking actions, what about `KernelStaker`?

File: kernel/contracts/ZepCore.sol [hunk 5]
```diff
@@ -3,77 +3,60 @@ pragma solidity ^0.4.18;
...
 
     uint256 stakedAmount = amount.sub(developerPayout);
-    _totalStaked = totalStaked().add(stakedAmount);
-    _instanceVouches[instance] = totalStakedFor(instance).add(stakedAmount);
-    _userVouches[msg.sender][instance] = _userVouches[msg.sender][instance].add(stakedAmount);
-    token().transfer(instance.developer(), developerPayout);
-
-    Staked(msg.sender, instance, stakedAmount, totalStakedFor(instance), data);
+    _stakes.stake(msg.sender, instance, stakedAmount, data);
+    _token.transfer(instance.developer(), developerPayout);
```
Comment (facuspagnuolo) [inline kernel/contracts/ZepCore.sol]: what if we move the `transferFrom` just before we stake them (L49)? Just following fail-fast pattern

File: kernel/contracts/ZepCore.sol (hunk 1 above)
Comment (spalladino) [reply kernel/contracts/ZepCore.sol]: This is why code reviews are needed. Thanks!

File: kernel/contracts/KernelStakes.sol (hunk 2 above)
Comment (spalladino) [reply kernel/contracts/KernelStakes.sol]: Yup, I embarrasingly missed it. Pushing a fix now!

File: kernel/contracts/KernelStakes.sol (hunk 3 above)
Comment (spalladino) [reply kernel/contracts/KernelStakes.sol]: :-)

File: kernel/contracts/KernelStakes.sol (hunk 4 above)
Comment (spalladino) [reply kernel/contracts/KernelStakes.sol]: Hmm I'd understand the "staker" to be "the person who stakes", not the class that manages it

File: kernel/contracts/ZepCore.sol (hunk 5 above)
Comment (spalladino) [reply kernel/contracts/ZepCore.sol]: Not sure I follow. `transferFrom` is the first call in the function.

File: kernel/contracts/KernelStakes.sol (hunk 2 above)
Comment (fiiiu) [reply kernel/contracts/KernelStakes.sol]: I would remove this method altogether now, since it is no longer called from `ZepCore`'s `transferStake`...

**Pull request merged**